src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
+            } else if current_fragment.has_special_character() {
+                let partial_glob_matcher =
+                    GlobBuilder::new(GlobFragment::join(&glob_fragments[..i + 1], "/").raw_str())
+                        .literal_separator(true)
+                        .build()
+                        .expect("Cannot parse glob")
+                        .compile_matcher();
+                let next_level_file_metadata =
+                    source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                match next_level_file_metadata {
+                    Ok(mut next_level_file_metadata) => {
+                        while let Some(fm) = next_level_file_metadata.next().await {
+                            match fm {
+                                Ok(fm) => {
+                                    if matches!(fm.filetype, FileType::Directory) && partial_glob_matcher.is_match(fm.filepath.as_str().trim_end_matches('/')) {
+                                        visit(
+                                            result_tx.clone(),
+                                            source.clone(),
+                                            fm.filepath.as_str(),
+                                            (glob_fragments.clone(), i + 1),
+                                            full_glob_matcher.clone(),
+                                        );
+                                    }
+                                }
+                                Err(super::Error::NotFound { .. }) => (),
+                                Err(e) => result_tx.send(Err(e)).await.expect("Internal multithreading channel is broken: results may be incorrect"),
+                            }
+                        }
+                    }
+                    Err(e) => result_tx.send(Err(e)).await.expect(
+                        "Internal multithreading channel is broken: results may be incorrect",
+                    ),
+                }
+
+            // RECURSIVE CASE: current_fragment contains no special characters, and is a path to a specific File or Directory
+            } else {
+                let full_dir_path = path.to_string() + current_fragment.escaped_str().as_str();
+                visit(
+                    result_tx.clone(),
+                    source.clone(),
+                    full_dir_path.as_str(),
+                    (glob_fragments.clone(), i + 1),
+                    full_glob_matcher.clone(),
+                );
+            }
+        });
+    }
+
+    visit(
+        to_rtn_tx,
+        source.clone(),
+        "",
+        (Arc::new(glob_fragments), 0),
+        Arc::new(full_glob_matcher),
+    );
+
+    let to_rtn_stream = stream! {
+        while let Some(v) = to_rtn_rx.recv().await {
+            // Glob only returns FileType::File results
+            assert!(v.as_ref().map_or(true, |v| matches!(v.filetype, FileType::File)));
```

Comment:
yield an error instead here

Reply:
This is more of a programmatic assert that I should only be getting File results, rather than an issue happening: maybe I can convert it into a filter to `continue` if it is not a `FileType::File`?

Reply:
I ended up just removing it